File: packages/checks/src/checks/noExplicitAnyReturn.ts
import { Array } from "effect"
import * as ts from "typescript"
import {
  hasAnyReturnType,
  isReturnTypeDeclaration,
  returnTypeDeclarationKinds
} from "./support/tsNode.js"
import type { CheckContext } from "@better-typescript/core/engine/check/data"
import type { Detection } from "@better-typescript/core/engine/location/data"

import { makeCheck } from "../defineCheck.js"
import { makeDetection } from "@better-typescript/core/engine/check"

const explicitAnyReturnElements = (context: CheckContext) => {
  const element = makeDetection(context)

  const matches = (node: ts.Node): ReadonlyArray<Detection> => {
    const reported = element({
      node,
      message: "Avoid function return types that include any.",
      hint:
        "Declare a precise return type instead of any. If the value is unknown at a boundary, " +
        "use unknown and narrow before use."
    })

    return isReturnTypeDeclaration(node) && hasAnyReturnType(node)
      ? Array.of(reported)
      : Array.empty()
  }

  return matches
}

export const noExplicitAnyReturn = makeCheck(
  "no-explicit-any-return",
  returnTypeDeclarationKinds,
  isReturnTypeDeclaration,
  explicitAnyReturnElements
)

File: internal/rules/no_explicit_any_return/no_explicit_any_return.go
package no_explicit_any_return

import (
	"github.com/andrueandersoncs/better-typescript/internal/rule"
	"github.com/andrueandersoncs/typescript-go/ast"
)

var message = rule.RuleMessage{
	Id:          "noExplicitAnyReturn",
	Description: "Avoid function return types that include any.",
	Help:        "Declare a precise return type instead of any. If the value is unknown at a boundary, use unknown and narrow before use.",
}

func containsAnyKeyword(node *ast.Node) bool {
	if node.Kind == ast.KindAnyKeyword {
		return true
	}
	for child := range node.IterChildren() {
		if containsAnyKeyword(child) {
			return true
		}
	}
	return false
}

func checkReturnType(ctx rule.RuleContext, node *ast.Node) {
	typeNode := node.Type()
	if typeNode != nil && containsAnyKeyword(typeNode) {
		ctx.ReportNode(node, message)
	}
}

var NoExplicitAnyReturnRule = rule.Rule{
	Name: "no-explicit-any-return",
	Run: func(ctx rule.RuleContext, options any) rule.RuleListeners {
		return rule.RuleListeners{
			ast.KindFunctionDeclaration: checkReturnTypeListener(ctx),
			ast.KindFunctionExpression:  checkReturnTypeListener(ctx),
			ast.KindArrowFunction:       checkReturnTypeListener(ctx),
			ast.KindMethodDeclaration:   checkReturnTypeListener(ctx),
			ast.KindMethodSignature:     checkReturnTypeListener(ctx),
			ast.KindCallSignature:       checkReturnTypeListener(ctx),
			ast.KindFunctionType:        checkReturnTypeListener(ctx),
			ast.KindGetAccessor:         checkReturnTypeListener(ctx),
		}
	},
}

func checkReturnTypeListener(ctx rule.RuleContext) func(*ast.Node) {
	return func(node *ast.Node) { checkReturnType(ctx, node) }
}
